From 8e951d2d0c63696e8c06eaa9b0225ec6b262b920 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Tue, 17 Aug 2010 01:27:31 -0700
Subject: [PATCH] wrhv/dpa: fix the dpa_bus_to_virt bug when PAMU is enabled

When PAMU is disabled, the physical address is equal to dma
address in guest OS. But when PAMU is enabled, maybe the
physical address and dma address are different. And we have
to use the dma address of PAGE_OFFSET to get the correct
virtual address.

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/dpa/dpa.c |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/net/dpa/dpa.c b/drivers/net/dpa/dpa.c
index 3a9700e..3b3bdd1 100644
--- a/drivers/net/dpa/dpa.c
+++ b/drivers/net/dpa/dpa.c
@@ -94,10 +94,10 @@ static const char rtx[][3] = {
 };
 
 #ifdef	CONFIG_PPC85xx_VT_MODE
-static u64 guest_phy_offset;
+static u64 guest_dma_offset;
 
 #define dpa_bus_to_virt(addr) 	\
-	__va(addr - guest_phy_offset)
+	__va(addr - guest_dma_offset)
 #else
 #define	dpa_bus_to_virt(addr)	\
 	bus_to_virt(addr)
@@ -2436,9 +2436,9 @@ static int __init __cold dpa_load(void)
 		goto _return;
 	}
 #endif
-
 #ifdef	CONFIG_PPC85xx_VT_MODE
-	vbi_guest_phys_to_phys((void *)virt_to_phys((void *)PAGE_OFFSET), &guest_phy_offset);
+	vbi_get_guest_dma_addr((void *)virt_to_phys((void *)PAGE_OFFSET),
+				&guest_dma_offset);
 #endif
 	_errno = of_register_platform_driver(&dpa_driver);
 	if (unlikely(_errno < 0)) {
-- 
1.6.5.2

